src/lib/uptime_service/uptime_service.ml

```diff
@@ -75,60 +75,101 @@ let send_uptime_data ~logger ~interruptor ~(submitter_keypair : Keypair.t) ~url
+                        more retries"
+                       max_attempts ~metadata
+                   else
+                     let base_metadata =
                        [ ("state_hash", State_hash.to_yojson state_hash)
-                       ; ( "includes_snark_work"
-                         , `Bool (Option.is_some block_data.snark_work) )
-                       ; ("is_produced_block", `Bool produced)
-                       ; ("url", `String (Uri.to_string url)) ]
-                 else
-                   let base_metadata =
+                       ; ("url", `String (Uri.to_string url))
+                       ; ("http_code", `Int (Cohttp.Code.code_of_status status))
+                       ; ( "http_error"
+                         , `String (Cohttp.Code.string_of_status status) ) ]
+                     in
+                     let extra_metadata = metadata_of_body body in
+                     let metadata = base_metadata @ extra_metadata in
+                     [%log info]
+                       "Failure when sending block with state hash \
+                        $state_hash to uptime service at URL $url, attempt %d \
+                        of %d, retrying"
+                       attempt max_attempts ~metadata ) ;
+                   succeeded
+             in
+             match%map.Deferred.Let_syntax
+               Interruptible.force interruptible
+             with
+             | Ok succeeded ->
+                 succeeded
+             | Error _ ->
+                 [%log error]
+                   "In uptime service, POST of block with state hash \
+                    $state_hash was interrupted"
+                   ~metadata:
                      [ ("state_hash", State_hash.to_yojson state_hash)
-                     ; ("url", `String (Uri.to_string url))
-                     ; ("http_code", `Int (Cohttp.Code.code_of_status status))
-                     ; ( "http_error"
-                       , `String (Cohttp.Code.string_of_status status) )
-                     ; ("payload", json) ]
-                   in
-                   let extra_metadata =
-                     match body with
-                     | `String s ->
-                         [("error", `String s)]
-                     | `Strings ss ->
-                         [ ( "error"
-                           , `List (List.map ss ~f:(fun s -> `String s)) ) ]
-                     | `Empty | `Pipe _ ->
-                         []
-                   in
-                   let metadata = base_metadata @ extra_metadata in
-                   [%log error]
-                     "Failure when sending block with state hash $state_hash \
-                      to uptime service at URL $url"
-                     ~metadata
+                     ; ("payload", json) ] ;
+                 (* interrupted, don't want to retry, claim success *)
+                 true
+           in
+           let rec go attempt =
```

Comment:
Nit: it would be clearer to me if this were folded up into `run_attempt`, so the failure case does the recursion.